File: safety_lab/src/safety_node.cpp
#include <ros/ros.h>
#include <sensor_msgs/LaserScan.h>
#include <nav_msgs/Odometry.h>
#include <ackermann_msgs/AckermannDriveStamped.h>
#include <std_msgs/Bool.h>

class Safety
{
    public:
        Safety(ros::NodeHandle &nh): rate_(120)    
        {
            nh_ = nh;
            relative_speed_ = 0.0;
            ttc_threshold_ = 0.20;
        }

        ~Safety(){}
        
        void initialize(){

            brake_pub_ = nh_.advertise<ackermann_msgs::AckermannDriveStamped>("/brake", 1);
            brake_bool_pub_ = nh_.advertise<std_msgs::Bool>("/brake_bool", 1);

            odom_sub_ = nh_.subscribe("/odom", 1, &Safety::odometryCallback, this);
            scan_sub_ = nh_.subscribe("/scan", 1, &Safety::laserScanCallback, this);
        }

        // Functions
        void odometryCallback(const nav_msgs::Odometry::ConstPtr& odom_msg){
            relative_speed_ = odom_msg->twist.twist.linear.x;
        }

        void laserScanCallback(const sensor_msgs::LaserScan::ConstPtr& scan_msg){
            
            auto current_angle = scan_msg->angle_min;
            const auto increment = scan_msg->angle_increment;
            double min_ttc = std::numeric_limits<double>::max();

            for (int i=0; i < scan_msg->ranges.size(); i++)
            {
                if(!std::isinf(scan_msg->ranges[i]) && !std::isnan(scan_msg->ranges[i]))
                {
                    const double current_ttc = scan_msg->ranges[i] / std::max(0.0, relative_speed_ * cos(current_angle));

//                    std::cout << current_ttc << "\n";
                    if (current_ttc < min_ttc)
                    {
                        min_ttc = current_ttc;
                    }
                }

                current_angle += increment;
            }

            if (min_ttc < ttc_threshold_)
            {
                ackermann_msgs::AckermannDriveStamped brake_msg;
                brake_msg.header = scan_msg->header;
                brake_msg.drive.speed = 0.0;

                std_msgs::Bool brake_bool_msg;
                brake_bool_msg.data = true;

                brake_bool_pub_.publish(brake_bool_msg);
                brake_pub_.publish(brake_msg);
            }
            else
            {
                std_msgs::Bool brake_bool_msg;
                brake_bool_msg.data = false;
                brake_bool_pub_.publish(brake_bool_msg);
            }

            rate_.sleep();
        }

    private:
        ros::NodeHandle nh_;

        // Publishers
        ros::Publisher brake_pub_;
        ros::Publisher brake_bool_pub_;

        // Subscribers
        ros::Subscriber odom_sub_;
        ros::Subscriber scan_sub_;

        // Callback rate
        ros::Rate rate_;

        // Other variables
        double relative_speed_;
        double ttc_threshold_;

};

int main(int argc, char **argv){

    ros::init(argc, argv, "AEB_node");

    ros::NodeHandle nh;

    Safety sf(nh);
    sf.initialize();

    ros::spin();

    return 0;
}

File: auto_emergency_braking/scripts/safety_node.py
#!/usr/bin/env python
import rospy
import math

from std_msgs.msg import Bool
from sensor_msgs.msg import LaserScan
from nav_msgs.msg import Odometry
from ackermann_msgs.msg import AckermannDriveStamped as ADS

class Safety():
    def __init__(self):

        rospy.init_node("safety_node", anonymous=True)
        self.brake_pub = rospy.Publisher("/brake", ADS, queue_size=1)
        self.brake_bool_pub = rospy.Publisher("/brake_bool", Bool, queue_size=1)

        rospy.Subscriber("/scan", LaserScan, self.scan_callback)
        rospy.Subscriber("/odom", Odometry, self.odom_callback)

        self.brake_bool = False

        self.brake_msg = ADS()
        self.speed = 0
        self.t_min = 0.25

    def odom_callback(self, odom_msg):

        self.speed = odom_msg.twist.twist.linear.x

    def scan_callback(self, scan_msg):

        ttc_min = float('inf')
        scan_ranges = scan_msg.ranges
        angle_min = scan_msg.angle_min
        increment = scan_msg.angle_increment

        current_angle = angle_min

        for i in range(len(scan_ranges)):

            if not math.isnan(scan_ranges[i]) and not math.isinf(scan_ranges[i]) and self.speed:
                r_dot = -self.speed*math.cos(current_angle)

                max_r_dot = max(-r_dot, 0)

                current_ttc = scan_ranges[i]/max_r_dot if max_r_dot else float('inf')

#                if current_angle < 0.025 and current_angle > -0.025:
#                    print(current_ttc)
#
                if current_ttc < ttc_min:
                    ttc_min = current_ttc
#
            current_angle = current_angle + increment
#
        print(ttc_min)

#        if ttc_min < self.t_min:
#            print("Inside loop")
#            #self.brake_msg.drive.steering_angle = 0
#            #self.brake_msg.drive.steering_angle_velocity = 0
#            self.brake_msg.header = scan_msg.header
#            self.brake_msg.drive.speed = 0
#
#            self.brake_bool = True
#            self.brake_bool_pub.publish(self.brake_bool)
#            self.brake_pub.publish(self.brake_msg)
#        else:
#            self.brake_bool = False
#            self.brake_bool_pub.publish(self.brake_bool)


if __name__ == "__main__":
    safe = Safety()

    rospy.Rate(100)

    rospy.spin()
